### ISSUE TITLE
Remmina Next is Corrupting the Unity Desktop

### ISSUE THREAD

**Issue by Lonniebiz** [FreeRDP/Remmina#1111]
For the past month or so, I've been having problems with Ubuntu's Unity desktop while running Remmina from the Next repository.

In Ubuntu 16.04 and in 16.10, maximizing the Remmina remote desktop window successfully maximizes the first time you do it (meaning its title bar does integrate into the top panel like it is suppose to in Unity).

However, if you click the Remmina-launcher (from Ubuntu Unitity's left launcher panel), to switch between Remmina's open-windows, it loses this top-panel-title-bar-integration and instead maximizes where the title bar is right below the top-panel  -- effectively reducing the vertical-size of your  remote desktop's screen -- and it will not maximize correctly until you close Remmina and then reopen all your remote desktops again.

In Ubuntu 16.10, while using Remmina next (maximized), I sometimes cannot access the menus of my applet-indicators that are located on Unity's  top-panel by the clock. I will click on the network-manager-applet (for example), expecting a menu to drop down, but because I have Remmina running the menu will not appear. As soon as I quit Remmina, the menus will start functioning again.

The bottom line is, about a month ago, Remmina-next stop playing nicely with the Ubuntu Unity desktop. Please figure out what code changes happened around that time, and just undo those changes. Please test against the Unity Desktop to ensure Remmina can maintain "title bar integration into the top panel during Maximization" while "clicking the remmina launcher to switch between open Remmina windows".

In Ubuntu 16.10, please make sure that Remmina doesn't somehow (invisibly) layer-over the application-indicators in the top panel rendering their menus unaccessible.

**Comment by giox069** [comment]
@Lonniebiz I'm already investigating on a very similar problem, and I think I'm affected too. Are you using keyboard capture ? (the keyboard capture button on the toolbar is pressed ?).
Let us know, and if possible, try a quick test without keyboard capture. Thank you.

**Comment by Lonniebiz** [comment]
Yes, I almost always use the keyboard capture feature for all my remote desktops. However, I will turn that off long enough to report back some results.

**Comment by Lonniebiz** [comment]
I think you're right about this issue having to do with code-changes that took place within the keyboard capture functionality; I'm unable to reproduce this when all of the remote desktops (I have open) are NOT using "keyboard capture".

The workaround I've been doing is, instead of clicking the Remmina-Launcher, I scroll my mouse-wheel while hovering over it (to switch between open Remmina windows), but when I forget to do this, I have to close Remmina and reOpen all my remote-desktops.

**Comment by Lonniebiz** [comment]
This bug completely corrupts the Unity desktop after it occurs. Even after you close Remmina, your desktop remains corrupted until you reboot your machine. This is hard to explain, so I just show you a screen-cast of what I'm talking about. Watch this:
[https://youtu.be/Za9Cjmq-9MY](https://youtu.be/Za9Cjmq-9MY)

**Comment by giox069** [comment]
In the video I can see the same effects I have on my desktop. When the desktop is in "corrupted state", appindicator icons becomes white. Sometimes compiz crashes. The only way to get out of this state is to disable keyboard capture, repeatdly change application and the click on the remmina icon on the launcher. Randomly, after 6-7 times, your desktop becomer responsive again.
It's not remmina fault, but a problem with unity or compiz: sometimes compiz crashestoo. I have just published a patch, but it totally breaks Gnome Shell capture. Event capture is a hell across different desktops. I'm still working on it.

**Comment by Lonniebiz** [comment]
When you say "it totally breaks Gnome Shell capture", are you saying that the "Grab all keyboard events" feature is broken in your current patch?

That feature was working great about a month ago. Are you saying that something must have changed in Compiz or Unity that ultimately caused this breakage in Remmina recently? If so, should we report a bug on launchpad.net regarding compiz and unity, in light of the current behavior we a witnessing in Remmina?

**Comment by Lonniebiz** [comment]
I made another screencast showcasing the behavior where application-indicator's context menus are not appearing, while running Remmina. I'm experiencing this in Ubuntu 16.10, and I don't recall having experienced this in 16.04:

https://youtu.be/pqlONEhIpAs

**Comment by antenore** [comment]
@Lonniebiz did you try the last commit or did you just install the package from the PPA? Giovanni should have fixed it, but to test it you should compile remmina manually or wait in 1 oer 2 days when we will release a new version (1.2.0-rcgit-18).

**Comment by 3v1n0** [comment]
Unity dev here... So I might help.

To fix the state of your desktop, I think that what you need to do is just to hit `Super+W` or just `Super` twice, in order to go out from the "state" mode...

I'll try to debug this in unity, but in general what should be done by such apps, is to grab the mouse, only when the pointer is covering the window surface.

**Comment by giox069** [comment]
@3v1n0 thanks for helping. Indeed, after a lot of trial-and-errror, I think I came to the same conclusion: grabbing should be released before leaving the window content area.
So a couple of days ago I started listening to GTK **leave-notify-event**, as older version of remmina did.
[Here](https://github.com/FreeRDP/Remmina/blob/8b4b67c223d4f5961f23d17270cb3c4dd5547a58/remmina/src/remmina_connection_window.c#L2156) you can see the latest patches in response to the GTK **leave-notify-event**.
As you can see, it has not been very simple to understand when **leave-notify-event** is really telling us that the pointer is leaving our remmina_connection_window. Under unity it was very simple: my window receive only one **leave-notify-event** when the pointer runs outside. Under gnome shell I received a lot of **leave-notify-event** even when the mouse moved from one widget to another inside the window. Under XFCE extra attention was needed to release the keyboard grab even when the mouse was on the window title bar.
These patches will go online tomorrow, but If you have other suggestions, we will be happy to hear :)

**Comment by 3v1n0** [comment]
One thing you can do to handle events when there's something grabbing them is also using xinput2, which basically allows you to get events anyway...

This is [what we do in unity](http://bazaar.launchpad.net/~unity-team/unity/trunk/view/head:/unity-shared/InputMonitor.cpp), as reference.

**Comment by antenore** [comment]
Thanks @3v1n0 isn't this something we should avoid as Wayland is "deprecating" the  X11part?

**Comment by Lonniebiz** [comment]
@3v1n0 Thanks for the Super+W suggestion. It has been repairing things for me and it is also a nice feature I was unaware of. Also thanks to @antenore and @giox069 ;  you guys are aways making it happen with Remmina.

**Comment by 3v1n0** [comment]
@antenore well, there's also mir coming :-).

Anyway, well... X11 will be there for long time, so I guess you still want to support it.
Also as a backup solution for Wayland/Mir (through X{Mir,Wayland} backends)

**Comment by Lonniebiz** [comment]
I'm just checking back on this. It has been a few days, and I haven't seen the update come through the Remmina-Next repository during typical Ubuntu Updates I've done.

**Comment by 3v1n0** [comment]
@Lonniebiz if you want, you can try the latest version of remmina also by using `snap`, see [the wiki](https://github.com/FreeRDP/Remmina/wiki#for-users-with-a-distro-that-supports-snap-packages-including-ubuntu-and-debian), but to get latest git version you should just do `sudo snap install remmina --edge`.

**Comment by giox069** [comment]
Yes, the remmina-next PPA is not yet updated. I hope @nastasi  will fix it ASAP. In the meanwhile you could test the snap package as suggested by @3v1n0 .

**Comment by giox069** [comment]
Remmina 1.2.0-rcgit-18 is now on the remmina-next PPA

**Comment by Lonniebiz** [comment]
Yes, I worked using it yesterday and all night (last night). So far, it is rock-solid; the bug has NOT occurred (in any form). I'm thankful; I know this one wasn't easy.

**Comment by HuboAdmins** [comment]
Same here, appears to be resolved in Remmina 1.2.0-rcgit-18. Appreciate it!

**Comment by giox069** [comment]
Thank you, I'm closing this issue.
